File: Device/SFBAudioControl.m
/*
 * Copyright (c) 2020 Stephen F. Booth <me@sbooth.org>
 * See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
 */

#import "SFBAudioControl.h"
#import "SFBAudioObject+Internal.h"

@implementation SFBAudioControl

- (instancetype)initWithAudioObjectID:(AudioObjectID)objectID
{
	NSParameterAssert(SFBAudioObjectIsClassOrSubclassOf(objectID, kAudioControlClassID));
	return [super initWithAudioObjectID:objectID];
}

- (NSNumber *)scope
{
	return [self unsignedIntForProperty:kAudioControlPropertyScope inScope:SFBAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSNumber *)element
{
	return [self unsignedIntForProperty:kAudioControlPropertyElement inScope:SFBAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

@end

File: Device/SFBAudioControl.swift
/*
* Copyright (c) 2020 Stephen F. Booth <me@sbooth.org>
* See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
*/

import Foundation

extension AudioControl {
	/// Returns the control's scope
	/// - note: This corresponds to `kAudioControlPropertyScope`
	/// - throws: An error if the property could not be retrieved
	public func scope() throws -> PropertyScope {
		return AudioObject.PropertyScope(rawValue: try getProperty(.controlScope))!
	}

	/// Returns the control's element
	/// - note: This corresponds to `kAudioControlPropertyElement`
	/// - throws: An error if the property could not be retrieved
	public func element() throws -> PropertyElement {
		return AudioControl.PropertyElement(try getProperty(.controlElement) as UInt32)
	}
}
